File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/hub/HubViewModel.kt
@file:Suppress("MagicNumber", "ComplexCondition")

package app.pantopus.android.ui.screens.hub

import android.content.SharedPreferences
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.hub.HubResponse
import app.pantopus.android.data.api.models.hub.HubTodayResponse
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.hub.HubRepository
import app.pantopus.android.ui.components.IdentityPillar
import app.pantopus.android.ui.theme.PantopusIcon
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import javax.inject.Inject

/** Key used to persist the dismissal flag for the amber setup banner. */
private const val BANNER_DISMISSED_KEY = "hub.setupBanner.dismissed"

/** ViewModel backing the hub screen. */
@HiltViewModel
class HubViewModel
    @Inject
    constructor(
        private val repo: HubRepository,
        private val prefs: SharedPreferences,
    ) : ViewModel() {
        private val _state = MutableStateFlow<HubUiState>(HubUiState.Skeleton)

        /** Observed hub state. */
        val state: StateFlow<HubUiState> = _state.asStateFlow()

        /** Initial load; no-op when already populated. */
        fun load() {
            if (_state.value is HubUiState.Populated) return
            refresh()
        }

        /** Pull-to-refresh / retry. */
        fun refresh() {
            _state.value = HubUiState.Skeleton
            viewModelScope.launch { fetch() }
        }

        /** Persist the amber-banner dismissal and hide it from the current state. */
        fun dismissSetupBanner() {
            prefs.edit().putBoolean(BANNER_DISMISSED_KEY, true).apply()
            (_state.value as? HubUiState.Populated)?.let { current ->
                _state.value = HubUiState.Populated(current.content.copy(setupBanner = null))
            }
        }

        private suspend fun fetch() {
            val hubResult = repo.overview()
            val hub =
                when (hubResult) {
                    is NetworkResult.Success -> hubResult.data
                    is NetworkResult.Failure -> {
                        _state.value = HubUiState.Error(hubResult.error.message)
                        return
                    }
                }

            // Companion endpoints run in parallel *after* hub so test sequences
            // can predict stub consumption. Failures degrade gracefully.
            val (today, discovery) =
                coroutineScope {
                    val todayJob = async { (repo.today() as? NetworkResult.Success)?.data }
                    val discoveryJob = async { (repo.discovery() as? NetworkResult.Success)?.data }
                    todayJob.await() to discoveryJob.await()
                }

            applyResults(hub, today, discovery?.items.orEmpty())
        }

        private fun applyResults(
            hub: HubResponse,
            today: HubTodayResponse?,
            discoveryItems: List<app.pantopus.android.data.api.models.hub.DiscoveryItem>,
        ) {
            val todaySummary = projectToday(today)
            val identity = primaryIdentity(hub)
            val discoveryCards =
                discoveryItems.take(10).map {
                    val kind = DiscoveryKind.fromRawType(it.type)
                    DiscoveryCardContent(
                        id = it.id,
                        title = it.title,
                        meta = it.meta,
                        category = it.category.orEmpty(),
                        avatarInitials = initials(it.title),
                        kind = kind,
                        tint = tintForDiscoveryKind(kind),
                    )
                }
            if (isFirstRun(hub)) {
                _state.value = firstRunState(hub, identity, discoveryCards)
                return
            }

            val bannerDismissed = prefs.getBoolean(BANNER_DISMISSED_KEY, false)
            val setupBanner = if (!hub.setup.allDone && !bannerDismissed) SetupBannerContent() else null

            _state.value =
                HubUiState.Populated(
                    PopulatedContent(
                        topBar =
                            TopBarContent(
                                greeting = greeting(),
                                name = hub.user.firstName ?: hub.user.name,
                                avatarInitials = initials(hub.user.name),
                                identity = identity,
                                ringProgress =
                                    hub.setup.profileCompleteness.score
                                        .toFloat(),
                                unreadCount = hub.statusItems.size,
                            ),
                        actionChips =
                            listOf(
                                ActionChipContent(ActionChipContent.Kind.PostTask, "Post task", PantopusIcon.PlusCircle, active = true),
                                ActionChipContent(ActionChipContent.Kind.SnapAndSell, "Snap & sell", PantopusIcon.Camera, active = false),
                                ActionChipContent(ActionChipContent.Kind.ScanMail, "Scan mail", PantopusIcon.ScanLine, active = false),
                                ActionChipContent(ActionChipContent.Kind.AddHome, "Add home", PantopusIcon.Home, active = false),
                            ),
                        setupBanner = setupBanner,
                        today = todaySummary,
                        pillars = pillars(hub, setupMode = false),
                        discovery = discoveryCards,
                        jumpBackIn =
                            hub.jumpBackIn.take(2).mapIndexed { index, raw ->
                                JumpBackItem(
                                    id = raw.title,
                                    title = raw.title,
                                    icon = iconFromRaw(raw.icon),
                                    route = raw.route,
                                    tint = tintForRoute(raw.route),
                                    kicker = if (index == 0) "In progress" else "Draft",
                                )
                            },
                        activity =
                            hub.activity.take(3).map {
                                ActivityEntry(
                                    id = it.id,
                                    title = it.title,
                                    timeAgo = relative(it.at),
                                    icon = PantopusIcon.Bell,
                                    tint = pillarTint(it.pillar),
                                )
                            },
                    ),
                )
        }

        private fun firstRunState(
            hub: HubResponse,
            identity: IdentityPillar,
            discoveryCards: List<DiscoveryCardContent>,
        ): HubUiState.FirstRun {
            val steps =
                hub.setup.steps.map {
                    SetupStep(
                        id = it.key,
                        title = it.key.replace('_', ' ').replaceFirstChar { c -> c.uppercase() },
                        done = it.done,
                    )
                }
            val doneCount = steps.count { it.done }
            return HubUiState.FirstRun(
                FirstRunContent(
                    greeting = greeting(),
                    name = hub.user.firstName ?: hub.user.name,
                    avatarInitials = initials(hub.user.name),
                    identity = identity,
                    ringProgress =
                        hub.setup.profileCompleteness.score
                            .toFloat(),
                    profileCompleteness =
                        hub.setup.profileCompleteness.score
                            .toFloat(),
                    stepsDone = doneCount,
                    stepsTotal = steps.size,
                    steps = steps,
                    pillars = pillars(hub, setupMode = true),
                    discovery = discoveryCards,
                ),
            )
        }

        private fun isFirstRun(hub: HubResponse): Boolean =
            !hub.setup.allDone &&
                hub.setup.profileCompleteness.score < 0.5 &&
                hub.homes.isEmpty()

        private fun projectToday(response: HubTodayResponse?): TodaySummary? {
            val today = response?.today ?: return null
            val weather = (today["weather"] as? Map<*, *>)
            val temperature = (weather?.get("temperatureF") as? Number)?.toInt()
            val conditions = weather?.get("conditions") as? String
            val aqi = ((today["aqi"] as? Map<*, *>)?.get("label") as? String)
            val commute = ((today["commute"] as? Map<*, *>)?.get("label") as? String)
            if (temperature == null && conditions == null && aqi == null && commute == null) return null
            return TodaySummary(
                temperatureFahrenheit = temperature,
                conditions = conditions,
                aqiLabel = aqi,
                commuteLabel = commute,
            )
        }

        private fun pillars(
            hub: HubResponse,
            setupMode: Boolean,
        ): List<PillarTile> {
            val personal = hub.cards.personal
            val home = hub.cards.home
            val business = hub.cards.business

            fun tile(
                kind: PillarTile.Pillar,
                label: String,
                icon: PantopusIcon,
                tint: IdentityPillar,
                chip: String?,
                chipSetupState: Boolean,
                populatedCaption: String,
                setupCaption: String,
            ) = PillarTile(
                pillar = kind,
                label = label,
                icon = icon,
                tint = tint,
                chip = if (setupMode) "Set up" else chip,
                chipSetupState = if (setupMode) true else chipSetupState,
                caption = if (setupMode) setupCaption else populatedCaption,
            )

            return listOf(
                tile(
                    PillarTile.Pillar.Pulse,
                    "Pulse",
                    PantopusIcon.Megaphone,
                    IdentityPillar.Personal,
                    if (personal.unreadChats > 0) "${personal.unreadChats} new" else null,
                    false,
                    if (personal.unreadChats > 0) "${personal.unreadChats} new in your feed" else "Neighborhood feed",
                    "Neighborhood feed",
                ),
                tile(
                    PillarTile.Pillar.Marketplace,
                    "Marketplace",
                    PantopusIcon.ShoppingBag,
                    IdentityPillar.Business,
                    if (business != null && business.newOrders > 0) "${business.newOrders}" else null,
                    business == null,
                    if (business != null) "${business.newOrders} new orders" else "Local buy & sell",
                    "Local buy & sell",
                ),
                tile(
                    PillarTile.Pillar.Gigs,
                    "Gigs",
                    PantopusIcon.Hammer,
                    IdentityPillar.Personal,
                    if (personal.gigsNearby > 0) "${personal.gigsNearby} matches" else null,
                    false,
                    if (personal.gigsNearby > 0) "${personal.gigsNearby} tasks near you" else "Earn & post tasks",
                    "Earn & post tasks",
                ),
                tile(
                    PillarTile.Pillar.Mail,
                    "Mail",
                    PantopusIcon.Mailbox,
                    IdentityPillar.Home,
                    if (home != null && home.newMail > 0) "${home.newMail}" else null,
                    home == null,
                    if (home != null) "${home.newMail} need pickup" else "Scan & forward",
                    "Scan & forward",
                ),
            )
        }

        /** Which identity tints the avatar ring. Defaults to home when
         *  the user has any claimed home; else personal. */
        private fun primaryIdentity(hub: HubResponse): IdentityPillar =
            if (hub.homes.isEmpty()) IdentityPillar.Personal else IdentityPillar.Home

        /** Maps a jump-back-in route to its pillar tint. */
        private fun tintForRoute(route: String): IdentityPillar =
            when {
                route.contains("gigs") || route.contains("post") -> IdentityPillar.Personal
                route.contains("marketplace") || route.contains("listings") -> IdentityPillar.Business
                route.contains("mail") || route.contains("homes") -> IdentityPillar.Home
                else -> IdentityPillar.Personal
            }

        private fun tintForDiscoveryKind(kind: DiscoveryKind): IdentityPillar =
            when (kind) {
                DiscoveryKind.Business -> IdentityPillar.Business
                else -> IdentityPillar.Personal
            }

        private fun pillarTint(value: String): IdentityPillar =
            when (value) {
                "home" -> IdentityPillar.Home
                "business" -> IdentityPillar.Business
                else -> IdentityPillar.Personal
            }

        private fun iconFromRaw(raw: String): PantopusIcon = PantopusIcon.valueOfRaw(raw) ?: PantopusIcon.ArrowLeft

        private fun initials(name: String): String =
            name
                .trim()
                .split(' ')
                .take(2)
                .mapNotNull { it.firstOrNull()?.uppercaseChar()?.toString() }
                .joinToString("")

        private fun greeting(): String {
            val hour = LocalDateTime.now(ZoneId.systemDefault()).hour
            return when (hour) {
                in 5..11 -> "Good morning"
                in 12..16 -> "Good afternoon"
                in 17..21 -> "Good evening"
                else -> "Hello"
            }
        }

        private fun relative(timestamp: String): String =
            runCatching {
                val then = Instant.parse(timestamp)
                val delta =
                    java.time.Duration
                        .between(then, Instant.now())
                        .seconds
                when {
                    delta < 60 -> "just now"
                    delta < 60 * 60 -> "${delta / 60}m ago"
                    delta < 60 * 60 * 24 -> "${delta / 60 / 60}h ago"
                    else -> "${delta / 60 / 60 / 24}d ago"
                }
            }.getOrDefault(timestamp)
    }

File: frontend/apps/ios/Pantopus/Features/Hub/HubViewModel.swift
//
//  HubViewModel.swift
//  Pantopus
//
//  Fetches `/api/hub`, `/api/hub/today`, and `/api/hub/discovery` in
//  parallel and projects them to the `HubState` consumed by `HubView`.
//

import Foundation
import Observation

/// ViewModel backing the designed hub screen.
@Observable
@MainActor
final class HubViewModel {
    /// Current state observed by `HubView`.
    private(set) var state: HubState = .skeleton

    private let api: APIClient
    private let bannerDismissedKey = "hub.setupBanner.dismissed"
    private var bannerDismissed: Bool {
        get { UserDefaults.standard.bool(forKey: bannerDismissedKey) }
        set { UserDefaults.standard.set(newValue, forKey: bannerDismissedKey) }
    }

    init(api: APIClient = .shared) {
        self.api = api
    }

    /// Initial load — no-op when we already have populated content.
    func load() async {
        if case .populated = state { return }
        state = .skeleton
        await fetch()
    }

    /// Pull-to-refresh / retry.
    func refresh() async {
        await fetch()
    }

    /// Dismiss the amber setup banner; persists across launches.
    func dismissSetupBanner() {
        bannerDismissed = true
        if case var .populated(content) = state {
            content = HubState.PopulatedContent(
                topBar: content.topBar,
                actionChips: content.actionChips,
                setupBanner: nil,
                today: content.today,
                pillars: content.pillars,
                discovery: content.discovery,
                jumpBackIn: content.jumpBackIn,
                activity: content.activity
            )
            state = .populated(content)
        }
    }

    // MARK: - Fetch

    private func fetch() async {
        // Hub first; its failure aborts the screen. Companions run in
        // parallel *after* a successful hub response so a test sequence
        // can predict which stub each call consumes.
        let hub: HubResponse
        do {
            hub = try await api.request(HubEndpoints.overview())
        } catch {
            state = .error(message: (error as? APIError)?.errorDescription ?? "Couldn't load your hub.")
            return
        }
        async let todayTask: HubTodayResponse? = optional {
            try await self.api.request(HubEndpoints.today())
        }
        async let discoveryTask: HubDiscoveryResponse? = optional {
            try await self.api.request(HubEndpoints.discovery(filter: "gigs", limit: 10))
        }
        let today = await todayTask
        let discovery = await discoveryTask
        applyResults(hub: hub, today: today, discovery: discovery)
    }

    /// Run a throwing async request and swallow its failure, returning nil.
    /// Used for the companion endpoints whose absence shouldn't block the hub.
    private func optional<T: Sendable>(_ operation: @Sendable () async throws -> T) async -> T? {
        try? await operation()
    }

    private func applyResults(
        hub: HubResponse,
        today: HubTodayResponse?,
        discovery: HubDiscoveryResponse?
    ) {
        let todaySummary = Self.projectToday(today)
        let identity = Self.primaryIdentity(for: hub)
        let isFirstRun = Self.isFirstRun(hub: hub)
        let discoveryCards = discovery?.items.prefix(10).map(Self.projectDiscovery(_:)) ?? []
        if isFirstRun {
            let steps = hub.setup.steps.map {
                SetupStep(id: $0.key, title: Self.setupTitle($0.key), done: $0.done)
            }
            let doneCount = steps.filter(\.done).count
            state = .firstRun(
                HubState.FirstRunContent(
                    greeting: Self.greeting(),
                    name: hub.user.firstName ?? hub.user.name,
                    avatarInitials: Self.initials(from: hub.user.name),
                    identity: identity,
                    ringProgress: hub.setup.profileCompleteness.score,
                    profileCompleteness: hub.setup.profileCompleteness.score,
                    stepsDone: doneCount,
                    stepsTotal: steps.count,
                    steps: steps,
                    // Setup-mode pillars + discovery rail, per the design's
                    // first-run frame.
                    pillars: Self.pillars(from: hub, setupMode: true),
                    discovery: discoveryCards
                )
            )
            return
        }

        let banner: SetupBannerContent? =
            (!hub.setup.allDone && !bannerDismissed)
                ? SetupBannerContent()
                : nil

        state = .populated(
            HubState.PopulatedContent(
                topBar: TopBarContent(
                    greeting: Self.greeting(),
                    name: hub.user.firstName ?? hub.user.name,
                    avatarInitials: Self.initials(from: hub.user.name),
                    identity: identity,
                    ringProgress: hub.setup.profileCompleteness.score,
                    unreadCount: hub.statusItems.count
                ),
                actionChips: Self.defaultActionChips(),
                setupBanner: banner,
                today: todaySummary,
                pillars: Self.pillars(from: hub, setupMode: false),
                discovery: discoveryCards,
                jumpBackIn: hub.jumpBackIn.prefix(2).enumerated().map { index, raw in
                    JumpBackItem(
                        id: raw.title,
                        title: raw.title,
                        icon: Self.icon(from: raw.icon),
                        route: raw.route,
                        tint: Self.tint(forRoute: raw.route),
                        // Backend doesn't carry kicker / progress for jump
                        // tiles yet — first slot reads "In progress",
                        // second reads "Draft" so the design's two-card
                        // visual lands without a backend change.
                        kicker: index == 0 ? "In progress" : "Draft",
                        progressLabel: nil,
                        progressFraction: nil
                    )
                },
                activity: hub.activity.prefix(3).map {
                    ActivityEntry(
                        id: $0.id,
                        title: $0.title,
                        timeAgo: Self.relative(timestamp: $0.at),
                        icon: .bell,
                        tint: Self.pillarTint(for: $0.pillar)
                    )
                }
            )
        )
    }

    // MARK: - Projections

    private static func isFirstRun(hub: HubResponse) -> Bool {
        !hub.setup.allDone
            && hub.setup.profileCompleteness.score < 0.5
            && hub.homes.isEmpty
    }

    private static func projectToday(_ response: HubTodayResponse?) -> TodaySummary? {
        guard let today = response?.today else { return nil }
        // Today is an opaque provider payload; best-effort extract common keys.
        let weather = today.dictValue?["weather"]?.dictValue
        let temperature = weather?["temperatureF"]?.numberValue.map { Int($0) }
        let conditions = weather?["conditions"]?.stringValue
        let aqiLabel = today.dictValue?["aqi"]?.dictValue?["label"]?.stringValue
        let commute = today.dictValue?["commute"]?.dictValue?["label"]?.stringValue
        if temperature == nil, conditions == nil, aqiLabel == nil, commute == nil { return nil }
        return TodaySummary(
            temperatureFahrenheit: temperature,
            conditions: conditions,
            aqiLabel: aqiLabel,
            commuteLabel: commute
        )
    }

    private static func projectDiscovery(_ item: HubDiscoveryResponse.Item) -> DiscoveryCardContent {
        let kind = DiscoveryKind(rawType: item.type)
        return DiscoveryCardContent(
            id: item.id,
            title: item.title,
            meta: item.meta,
            category: item.category ?? "",
            avatarInitials: initials(from: item.title),
            kind: kind,
            tint: tint(forDiscoveryKind: kind)
        )
    }

    private static func tint(forDiscoveryKind kind: DiscoveryKind) -> IdentityPillar {
        switch kind {
        case .business: .business
        case .person, .post, .gig, .unknown: .personal
        }
    }

    private static func pillars(from hub: HubResponse, setupMode: Bool) -> [PillarTile] {
        let personal = hub.cards.personal
        let home = hub.cards.home
        let business = hub.cards.business

        func pillar(
            _ kind: PillarTile.Pillar,
            label: String,
            icon: PantopusIcon,
            tint: IdentityPillar,
            chip: String?,
            chipSetupState: Bool,
            populatedCaption: String,
            setupCaption: String
        ) -> PillarTile {
            PillarTile(
                pillar: kind,
                label: label,
                icon: icon,
                tint: tint,
                chip: setupMode ? "Set up" : chip,
                chipSetupState: setupMode ? true : chipSetupState,
                caption: setupMode ? setupCaption : populatedCaption
            )
        }

        return [
            pillar(
                .pulse, label: "Pulse", icon: .megaphone, tint: .personal,
                chip: personal.unreadChats > 0 ? "\(personal.unreadChats) new" : nil,
                chipSetupState: false,
                populatedCaption: personal.unreadChats > 0
                    ? "\(personal.unreadChats) new in your feed"
                    : "Neighborhood feed",
                setupCaption: "Neighborhood feed"
            ),
            pillar(
                .marketplace, label: "Marketplace", icon: .shoppingBag, tint: .business,
                chip: business.map { $0.newOrders > 0 ? "\($0.newOrders)" : nil } ?? nil,
                chipSetupState: business == nil,
                populatedCaption: business.map { "\($0.newOrders) new orders" } ?? "Local buy & sell",
                setupCaption: "Local buy & sell"
            ),
            pillar(
                .gigs, label: "Gigs", icon: .hammer, tint: .personal,
                chip: personal.gigsNearby > 0 ? "\(personal.gigsNearby) matches" : nil,
                chipSetupState: false,
                populatedCaption: personal.gigsNearby > 0
                    ? "\(personal.gigsNearby) tasks near you"
                    : "Earn & post tasks",
                setupCaption: "Earn & post tasks"
            ),
            pillar(
                .mail, label: "Mail", icon: .mailbox, tint: .home,
                chip: home.map { $0.newMail > 0 ? "\($0.newMail)" : nil } ?? nil,
                chipSetupState: home == nil,
                populatedCaption: home.map { "\($0.newMail) need pickup" } ?? "Scan & forward",
                setupCaption: "Scan & forward"
            )
        ]
    }

    /// Maps a `/app/…` jump route to the pillar tint that owns it. Falls
    /// back to `.personal` for routes that don't carry a pillar (the
    /// design's default).
    private static func tint(forRoute route: String) -> IdentityPillar {
        switch true {
        case route.contains("gigs"), route.contains("post"): .personal
        case route.contains("marketplace"), route.contains("listings"): .business
        case route.contains("mail"), route.contains("homes"): .home
        default: .personal
        }
    }

    /// Which identity tints the avatar ring + section accents. Default to
    /// `.home` if the user has any home claimed, else `.personal` (matches
    /// the design's most-common populated frame).
    private static func primaryIdentity(for hub: HubResponse) -> IdentityPillar {
        hub.homes.isEmpty ? .personal : .home
    }

    private static func defaultActionChips() -> [ActionChipContent] {
        [
            ActionChipContent(kind: .postTask, label: "Post task", icon: .plusCircle, active: true),
            ActionChipContent(kind: .snapAndSell, label: "Snap & sell", icon: .camera, active: false),
            ActionChipContent(kind: .scanMail, label: "Scan mail", icon: .scanLine, active: false),
            ActionChipContent(kind: .addHome, label: "Add home", icon: .home, active: false)
        ]
    }

    private static func pillarTint(for value: String) -> IdentityPillar {
        switch value {
        case "personal": .personal
        case "home": .home
        case "business": .business
        default: .personal
        }
    }

    private static func icon(from raw: String) -> PantopusIcon {
        PantopusIcon.allCases.first { $0.rawValue == raw } ?? .arrowLeft
    }

    private static func initials(from name: String) -> String {
        let parts = name.split(separator: " ").prefix(2)
        return parts.compactMap { $0.first.map(String.init) }.joined().uppercased()
    }

    private static func setupTitle(_ key: String) -> String {
        key.replacingOccurrences(of: "_", with: " ").capitalized
    }

    private static func greeting() -> String {
        let hour = Calendar.current.component(.hour, from: Date())
        switch hour {
        case 5..<12: return "Good morning"
        case 12..<17: return "Good afternoon"
        case 17..<22: return "Good evening"
        default: return "Hello"
        }
    }

    private static func relative(timestamp: String) -> String {
        guard let date = ISO8601DateFormatter().date(from: timestamp) else { return timestamp }
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .short
        return formatter.localizedString(for: date, relativeTo: Date())
    }
}

// `JSONValue.dictValue` lives in `Core/Networking/Models/Common/JSONValue.swift`.
